rama-net/src/client/pool.rs

```diff
@@ -0,0 +1,592 @@
+use super::conn::{ConnectorService, EstablishedClientConnection};
+use crate::stream::Socket;
+use parking_lot::Mutex;
+use rama_core::error::{BoxError, ErrorContext, OpaqueError};
+use rama_core::{Context, Service};
+use std::collections::VecDeque;
+use std::fmt::Debug;
+use std::num::NonZeroU16;
+use std::ops::{Deref, DerefMut};
+use std::pin::Pin;
+use std::sync::{Arc, Weak};
+use std::time::Duration;
+use std::{future::Future, net::SocketAddr};
+use tokio::io::{AsyncRead, AsyncWrite};
+use tokio::sync::{OwnedSemaphorePermit, Semaphore};
+use tokio::time::timeout;
+
+#[expect(dead_code)]
+#[derive(Debug)]
+/// Active slot is able to actively use a connection to make requests.
+/// They are used to track 'active' connections inside the pool
+struct ActiveSlot(OwnedSemaphorePermit);
+
+#[expect(dead_code)]
+#[derive(Debug)]
+/// Pool slot is needed to add a connection to the pool. Poolslots have
+/// a one to one mapping to connections inside the pool, and are used
+/// to track the 'total' connections inside the pool
+struct PoolSlot(OwnedSemaphorePermit);
+
+#[derive(Debug)]
+/// A connection which is stored in a pool. A hash is used to determine
+/// which connections can be used for a request. This hash encodes
+/// all the details that make a connection unique/suitable for a request.
+struct PooledConnection<C, ConnId> {
+    /// Actual raw connection that is stored in pool
+    conn: C,
+    /// ID is not unique but is used to group connections that can be used for the same request
+    id: ConnId,
+    /// Slot this connection takes up in the pool
+    slot: PoolSlot,
+}
+
+#[derive(Debug)]
```

Comment:
impl Debug for structs with generics manually please., impl only if all generics are debug. As this derive enforced that they have to be debug, which is a needless restriction.